sklearn/decomposition/dict_learning.py

```diff
@@ -343,35 +352,32 @@ def _update_dict(dictionary, Y, code, verbose=False, return_r2=False,
         Updated dictionary.
 
     """
-    n_components = len(code)
-    n_samples = Y.shape[0]
+    n_features, n_components = B.shape
     random_state = check_random_state(random_state)
     # Residuals, computed 'in-place' for efficiency
-    R = -np.dot(dictionary, code)
-    R += Y
+    R = -np.dot(dictionary, A)
+    R += B
     R = np.asfortranarray(R)
-    ger, = linalg.get_blas_funcs(('ger',), (dictionary, code))
+    ger, = linalg.get_blas_funcs(('ger',), (dictionary, A))
     for k in range(n_components):
         # R <- 1.0 * U_k * V_k^T + R
-        R = ger(1.0, dictionary[:, k], code[k, :], a=R, overwrite_a=True)
-        dictionary[:, k] = np.dot(R, code[k, :].T)
+        R = ger(1.0, dictionary[:, k], A[k, :], a=R, overwrite_a=True)
+        dictionary[:, k] = np.dot(R, A[k, :])
```

Comment:
even with n_iter=10000 and tol=1e-20 a difference persists. I computed the residual by hand with

```
import numpy as np
rng = np.random.RandomState(0)
X = rng.randn(30, 20)
U, V, _ = dict_learning(X, n_components=3, alpha=0.1, random_state=0, verbose=2, max_iter=10000, tol=0.00001)
print(0.5 * np.sum((X - np.dot(U, V)) ** 2) + 0.1 * np.sum(np.abs(U)))
```

On master I get 188.7666, and on the current branch I get 188.7859, so slightly more...